src/Controller/FragmentRegistry/FragmentRegistry.php

```diff
@@ -0,0 +1,153 @@
+<?php
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2016 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\Controller\FragmentRegistry;
+
+use Contao\CoreBundle\Controller\FragmentRegistry\FragmentType\FragmentInterface;
+use Symfony\Component\HttpKernel\Controller\ControllerReference;
+use Symfony\Component\HttpKernel\Fragment\FragmentHandler;
+
+/**
+ * Fragment registry.
+ *
+ * @author Yanick Witschi <https://github.com/toflar>
+ */
+class FragmentRegistry implements FragmentRegistryInterface
+{
+    /**
+     * @var FragmentHandler
+     */
+    private $fragmentHandler;
+
+    /**
+     * @var ControllerReference
+     */
+    private $controller;
+
+    /**
+     * @var array
+     */
+    private $types;
+
+    /**
+     * @var FragmentInterface[]
+     */
+    private $fragments = [];
+
+    /**
+     * @var FragmentInterface[]
+     */
+    private $fragmentsPerType = [];
+
+    /**
+     * @var FragmentInterface[]
+     */
+    private $fragmentsPerTypeAndName = [];
+
+    /**
+     * FragmentRegistry constructor.
+     *
+     * @param FragmentHandler $fragmentHandler
+     * @param string          $controllerName
+     */
+    public function __construct(FragmentHandler $fragmentHandler, $controllerName)
+    {
+        $this->fragmentHandler = $fragmentHandler;
+        $this->controller = $controllerName;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addFragmentType($interfaceClassName)
+    {
+        $this->types[] = $interfaceClassName;
```

Comment:
We should throw an exception here when fragments have already been registered as the new type will not have been evaluated for the existing ones otherwise.
The workflow must be:
1. register types
2. register fragments.

Other solution is to cycle through all existing fragments and evaluate against this type again.